lxd/instance/drivers/driver_qemu.go

```diff
@@ -1763,6 +1773,102 @@ func (d *qemu) deviceStart(deviceName string, rawConfig deviceConfig.Device, ins
 	return runConf, nil
 }
 
+func (d *qemu) deviceAttachBlockDevice(deviceName string, configCopy map[string]string, mounts []deviceConfig.MountEntryItem) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Check if the agent is running.
+	monitor, err := qmp.Connect(d.monitorPath(), qemuSerialChardevName, d.getMonitorEventHandler())
+	if err != nil {
+		return fmt.Errorf("Failed to connect to QMP monitor: %w", err)
+	}
+
+	fields := strings.Split(mounts[0].DevPath, ":")
+	blockFilename := fields[2]
+	oldDevPaths := d.devPaths
+	d.devPaths = append(d.devPaths, blockFilename)
+
+	err = apparmor.InstanceLoad(d.state.OS, d)
+	if err != nil {
+		return fmt.Errorf("Failed to load AppArmor policy: %w", err)
+	}
+
+	revert.Add(func() {
+		d.devPaths = oldDevPaths
+		apparmor.InstanceLoad(d.state.OS, d)
+	})
+
+	escapedDeviceName := filesystem.PathNameEncode(deviceName)
+
+	qemuDev := make(map[string]string)
+	qemuBlockDev := make(map[string]interface{})
+
+	qemuBlockDev["node-name"] = fmt.Sprintf("%s%s", qemuBlockDevIDPrefix, escapedDeviceName)
+	qemuBlockDev["driver"] = "raw"
+	qemuBlockDev["file"] = map[string]interface{}{
+		"driver":   "file",
+		"filename": blockFilename,
+	}
+
+	qemuDev["driver"] = "scsi-hd"
+	qemuDev["drive"] = qemuBlockDev["node-name"].(string)
+	qemuDev["id"] = fmt.Sprintf("%s%s", qemuDeviceIDPrefix, escapedDeviceName)
+
+	monHook := func(m *qmp.Monitor) error {
+		err := m.AddBlockDevice(qemuBlockDev, qemuDev)
+		if err != nil {
+			return fmt.Errorf("Failed to add block device: %w", err)
+		}
+
+		return nil
+	}
+
+	err = monHook(monitor)
+	if err != nil {
+		return fmt.Errorf("Failed to call monitor hook for block device: %w", err)
+	}
+
+	revert.Success()
+	return nil
+}
+
+func (d *qemu) deviceDetachBlockDevice(deviceName string, rawConfig deviceConfig.Device) error {
+	// Check if the agent is running.
+	monitor, err := qmp.Connect(d.monitorPath(), qemuSerialChardevName, d.getMonitorEventHandler())
+	if err != nil {
+		return err
+	}
+
+	escapedDeviceName := filesystem.PathNameEncode(deviceName)
+	deviceID := fmt.Sprintf("%s%s", qemuDeviceIDPrefix, escapedDeviceName)
+	blockDevName := fmt.Sprintf("%s%s", qemuBlockDevIDPrefix, escapedDeviceName)
+
+	err = monitor.RemoveBlockDevice(blockDevName, deviceID)
+	if err != nil {
+		return err
+	}
+
+	if rawConfig != nil {
+		var newDevPaths []string
+
+		for _, path := range d.devPaths {
```

Comment:
Same concern about d.devPaths here as above, if this is run when the VM is already running (i.e hot-unplug) won't we end up removing all previously allowed devPaths?

Reply:
We iterate through devPaths and only skip the entry of the disk which is to be removed. We don't remove all previously allowed devPaths.

Reply:
Yes i see that, but my understanding is that devPaths itself will be empty any other time than start up. So effectively this would remove start-up time devPaths from the apparmor profile.

Reply:
https://github.com/lxc/lxd/blob/57e0dbc26c2bde3ce2fc1ce33e5528fd16ef489b/lxd/instance/drivers/driver_common.go#L94-L97

Which makes sense, because we don't keep the `Instance` var instance around during the VM lifetime, a  new `Instance` var instance is created when updating a VM.